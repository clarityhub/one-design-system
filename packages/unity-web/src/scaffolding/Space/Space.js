import React from 'react';
import { css } from '@emotion/core';
import isPropValid from '@emotion/is-prop-valid';

import styled from '@emotion/styled';

import { oneOf } from 'prop-types';

import spacing from '@clarityhub/unity-core/lib/spacing';

const remMapper = (size) => {
	return `${spacing[size]}rem`;
};

const Space = styled(`div`, {
	shouldForwardProp: prop => isPropValid(prop) && ![
		'direction',
		'size',
	].includes(prop),
})(
	({ direction = 'horizontal', size = 'xsmall' }) => {
		const mappedSize = remMapper(size);
        
		if (direction === 'horizontal') {
			return css`
                > * + * {
                    margin-left: ${mappedSize}
                }
            `;
		} else {
			return css`
                > * + * {
                    margin-top: ${mappedSize}
                }
            `;
		}
	},
);

/**
 * Hack for docz and react-docgen
 */
const SpaceProps = () => <div />;
SpaceProps.propTypes = {
	direction: oneOf(['vertical', 'horizontal']),
	gap: oneOf([
		'xsmall',
		'small',
		'large',
		'xlarge',
		'medium',
	]),
};

Space.propTypes = SpaceProps.propTypes;
Space.defaultProps = SpaceProps.defaultProps;
export { SpaceProps };

export default Space;
