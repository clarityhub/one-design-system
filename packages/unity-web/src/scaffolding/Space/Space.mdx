---
name: Space
menu: Scaffolding
---

import { Playground, PropsTable } from 'docz';
import Space, { SpaceProps } from './Space';

# Space

Space is a primitive that you can use to space out children horizontally or vertically.

## \<Space /\> component

```js
import Space from '@clarityhub/unity-web/lib/scaffolding/Space';
```

## Examples

### Vertical Space

<Playground>
	<Space direction="vertical">
        <div><button>A</button></div>
        <div><button>B</button></div>
        <div><button>C</button></div>
    </Space>
</Playground>

### Vertical Space

<Playground>
	<Space direction="horizontal">
        <button style={{ display: 'inline-block' }}>A</button>
        <button style={{ display: 'inline-block' }}>B</button>
        <button style={{ display: 'inline-block' }}>C</button>
    </Space>
</Playground>

### Space Size

<Playground>
	<Space direction="vertical" size="large">
        <div><button>A</button></div>
        <div><button>B</button></div>
        <div><button>C</button></div>
    </Space>
</Playground>

## Prop Types

<PropsTable of={SpaceProps} />