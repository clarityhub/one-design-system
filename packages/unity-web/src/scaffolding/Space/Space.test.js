import React from 'react';
import renderer from 'react-test-renderer';

import Space from './Space';

describe('<Space />', () => {
	it('renders children', () => {
		const node = (
			<Space direction="vertical">
				<button>Test</button>
				<button>Test</button>
			</Space>
		);
		const tree = renderer.create(node).toJSON();

		expect(tree).toMatchSnapshot();
	});
});