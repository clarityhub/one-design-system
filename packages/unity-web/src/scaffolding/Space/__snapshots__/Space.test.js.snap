// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Space /> renders children 1`] = `
.emotion-0 > * + * {
  margin-top: 0.5rem;
}

<div
  className="emotion-0 emotion-1"
>
  <button>
    Test
  </button>
  <button>
    Test
  </button>
</div>
`;
